File: 5. Validations and transformations for backend engineers/code/typescript/example_3.ts
import { z } from 'zod';

// optional + then 7-15 digits (country code + national no.)
const phoneRe = /^\+?[0-9]{7,15}$/;

export const ContactSchema = z.object({
  email: z.string().email(), // local @ domain.tld
  phone: z.string().regex(phoneRe, "invalid phone number format"),
  date: z.string().regex(/^\d{4}-\d{2}-\d{2}$/, "only accepts YYYY-MM-DD") // YYYY-MM-DD
});
export type Contact = z.infer<typeof ContactSchema>;

File: 5. Validations and transformations for backend engineers/code/rust/example_3.rs
use serde::Deserialize;
use validator::Validate;
use regex::Regex;
use lazy_static::lazy_static;

lazy_static! {
    static ref PHONE_RE: Regex = Regex::new(r"^\+?[0-9]{7,15}$").unwrap();
}

#[derive(Debug, Deserialize, Validate)]
pub struct Contact {
    #[validate(email)]
    pub email: String, // local @ domain.tld
    #[validate(regex(path = "PHONE_RE", message = "invalid phone number format"))]
    pub phone: String, // checked below
    // custom validation for YYYY-MM-DD
    pub date: String, 
}
